pkg/agent/controller/l7visibility/l7visibility_controller_test.go

```diff
@@ -0,0 +1,546 @@
+	podInterface3 := newPodInterface(pod3.Name, pod3.Namespace, int32(pod3OFPort))
+	interfaces := []*interfacestore.InterfaceConfig{
+		podInterface1,
+		podInterface2,
+		podInterface3,
+	}
+	testcases := []struct {
+		name                      string
+		updatedPod                *v1.Pod
+		expectedPodToDirectionMap map[string]v1alpha2.Direction
+		expectedCalls             func(mockOFClient *openflowtest.MockClient)
+	}{
+		{
+			name:       "Update Pod with different annotation to correct annotation",
+			updatedPod: newPodObject("test-pod1", "test-ns1", annotationsCorrectEgress),
+			expectedPodToDirectionMap: map[string]v1alpha2.Direction{
+				pod1NN: v1alpha2.DirectionEgress,
+			},
+			expectedCalls: func(mockOFClient *openflowtest.MockClient) {
+				mockOFClient.EXPECT().InstallTrafficControlMarkFlows(fmt.Sprintf("tcl7:%s", pod1NN), []uint32{uint32(podInterface1.OFPort)}, targetPort, directionEgress, actionMirror, types.TrafficControlFlowPriorityLow)
+			},
+		}, {
+			name:       "Update Pod with Incorrect annotation to correct annotation",
+			updatedPod: newPodObject("test-pod2", "test-ns1", annotationsCorrectBoth),
+			expectedPodToDirectionMap: map[string]v1alpha2.Direction{
+				pod2NN: v1alpha2.DirectionBoth,
+			},
+			expectedCalls: func(mockOFClient *openflowtest.MockClient) {
+				mockOFClient.EXPECT().InstallTrafficControlMarkFlows(fmt.Sprintf("tcl7:%s", pod2NN), []uint32{uint32(podInterface2.OFPort)}, targetPort, directionBoth, actionMirror, types.TrafficControlFlowPriorityLow)
+			},
+		}, {
+			name:       "Update Pod with no annotation to correct annotation",
+			updatedPod: newPodObject("test-pod3", "test-ns1", annotationsCorrectIngress),
+			expectedPodToDirectionMap: map[string]v1alpha2.Direction{
+				pod3NN: v1alpha2.DirectionIngress,
+			},
+			expectedCalls: func(mockOFClient *openflowtest.MockClient) {
+				mockOFClient.EXPECT().InstallTrafficControlMarkFlows(fmt.Sprintf("tcl7:%s", pod3NN), []uint32{uint32(podInterface3.OFPort)}, targetPort, directionIngress, actionMirror, types.TrafficControlFlowPriorityLow)
+			},
+		},
+	}
+	for _, tt := range testcases {
+		t.Run(tt.name, func(t *testing.T) {
+			c := newFakeControllerAndWatcher(t, []runtime.Object{pod1, pod2, pod3, testNS1}, interfaces)
+			stopCh := make(chan struct{})
+			defer close(stopCh)
+
+			c.startInformers(stopCh)
+
+			assert.Eventuallyf(t, func() bool {
+				ns, _ := c.namespaceLister.List(labels.Everything())
+				return len(c.localPodInformer.GetIndexer().List()) == 3 && len(ns) == 1
+			}, 1*time.Second, 10*time.Millisecond, "Pods should be added to Informers")
+
+			// Ignore pods add events for pod2
+			waitEvents(t, 1, c)
+			item, _ := c.queue.Get()
+			c.queue.Done(item)
+
+			tt.expectedCalls(c.mockOFClient)
+
+			// Update Pods with new annotations
+			_, err := c.client.CoreV1().Pods(tt.updatedPod.Namespace).Update(context.TODO(), tt.updatedPod, metav1.UpdateOptions{})
```

Comment:
I prefer to have a `ctx := context.Background()` declaration at the beginning of the test, and use that throughout. `context.TODO()` implies (at least to me) that we should replace it at some point in the future, which is not the case here.